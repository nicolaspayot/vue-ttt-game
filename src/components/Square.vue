<template>
  <div
    class="game__box"
    :class="{
      'game__box--x': value === 'X',
      'game__box--o': value === 'O',
      'game__box--over': isGameOver,
      'game__box--winner': isWinner
    }"
    @click="onClick"
  >
    <span>{{ value }}</span>
  </div>
</template>

<script>
export default {
  props: {
    value: String,
    isGameOver: Boolean,
    isWinner: Boolean
  },
  methods: {
    onClick() {
      this.$emit("click");
    }
  }
};
</script>
