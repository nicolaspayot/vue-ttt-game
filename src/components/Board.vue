<template>
  <div class="game">
    <Status :winner="winner.player" :player="player" :isDrawGame="isDrawGame" />
    <div class="game__board">
      <Square
        v-for="(square, position) in squares"
        :key="position"
        :value="square"
        :isGameOver="isGameOver"
        :isWinner="isSquareWinner(position)"
        @click="onSquareClick(position)"
      />
    </div>
    <Restart @click="restartGame" />
  </div>
</template>

<script>
import getWinner from "../utils/winner";
import Status from "./Status";
import Square from "./Square";
import Restart from "./Restart";

export default {
  components: {
    Status,
    Square,
    Restart
  },
  data() {
    return {
      player: "X",
      squares: Array(9).fill(null)
    };
  },
  computed: {
    winner() {
      return getWinner(this.squares);
    },
    isDrawGame() {
      return this.squares.filter(val => val === null).length === 0;
    },
    isGameOver() {
      return this.winner.player !== null || this.isDrawGame;
    }
  },
  methods: {
    isSquareEmpty(position) {
      return this.squares[position] === null;
    },
    isSquareWinner(position) {
      return this.winner.positions.includes(position);
    },
    onSquareClick(position) {
      if (!this.isSquareEmpty(position) || this.winner.player) {
        return;
      }
      this.$set(this.squares, position, this.player);
      this.player = this.player === "X" ? "O" : "X";
    },
    restartGame() {
      this.player = "X";
      this.squares = Array(9).fill(null);
      this.$store.commit("clear");
      this.$router.push({ path: "/" });
    }
  }
};
</script>
