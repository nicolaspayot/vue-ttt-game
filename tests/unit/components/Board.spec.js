import { mount } from "@vue/test-utils";
import Board from "@/components/Board";
import Square from "@/components/Square";

describe("Board.vue", () => {
  let wrapper;
  let vm;

  beforeEach(() => {
    wrapper = mount(Board);
    vm = wrapper.vm;
  });

  it("should render a Board with 9 squares", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should be possible to play on an empty square when game is not over", async () => {
    // Given
    const squares = wrapper.findAll(Square);

    // When
    squares.at(0).trigger("click");

    // Then
    expect(vm.squares[0]).toBe("X");
    expect(vm.player).toBe("O");
  });

  it("should not be possible to play on a non empty square when game is not over", async () => {
    // Given
    vm.squares = ["X", "O", "X", null, null, null, null, null, null];
    vm.player = "O";
    await vm.$nextTick();

    const squares = wrapper.findAll(Square);

    // When
    squares.at(2).trigger("click");

    // Then
    expect(vm.squares[2]).toBe("X");
    expect(vm.player).toBe("O");
  });

  it("should not be possible to play when game has a winner", async () => {
    // Given
    vm.squares = ["X", "X", "X", "O", "O", null, null, null, null];
    vm.player = "O";
    await vm.$nextTick();

    const squares = wrapper.findAll(Square);

    // When
    squares.at(5).trigger("click");

    // Then
    expect(vm.squares[5]).toBe(null);
    expect(vm.player).toBe("O");
  });

  it("should display the next player when game is not over", async () => {
    // Given
    const squares = wrapper.findAll(Square);

    // When
    squares.at(0).trigger("click");
    squares.at(1).trigger("click");
    squares.at(2).trigger("click");
    await vm.$nextTick();

    // Then
    const status = wrapper.find(".game__status > span");
    expect(status.text()).toBe("Next player is O");
  });

  it("should display the winner", async () => {
    // Given
    vm.squares = ["O", "O", "O", "X", "X", null, null, null, null];
    await vm.$nextTick();

    // Then
    const status = wrapper.find(".game__status > span");
    expect(status.text()).toBe("Winner is O");
  });

  it("should display when game is a draw", async () => {
    // Given
    vm.squares = ["X", "O", "X", "O", "X", "X", "O", "X", "O"];
    await vm.$nextTick();

    // Then
    const status = wrapper.find(".game__status > span");
    expect(status.text()).toBe("That's a draw");
  });
});
