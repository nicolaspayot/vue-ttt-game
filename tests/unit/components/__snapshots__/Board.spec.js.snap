// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board.vue should render a Board with 9 squares 1`] = `
<div class="game">
  <div class="game__status"><span>Next player is X</span></div>
  <div class="game__board">
    <div class="game__box"><span></span></div>
    <div class="game__box"><span></span></div>
    <div class="game__box"><span></span></div>
    <div class="game__box"><span></span></div>
    <div class="game__box"><span></span></div>
    <div class="game__box"><span></span></div>
    <div class="game__box"><span></span></div>
    <div class="game__box"><span></span></div>
    <div class="game__box"><span></span></div>
  </div>
  <div class="game__action"><button>Restart game</button></div>
</div>
`;
